src/google/protobuf/compiler/parser_unittest.cc

```diff
@@ -509,6 +512,8 @@ TEST_F(ParseMessageTest, FieldDefaults) {
       " }"
       "  field { type:TYPE_DOUBLE  default_value:\"nan\"       " ETC
       " }"
+      "  field { type:TYPE_DOUBLE  default_value:\"-nan\"      " ETC
```

Comment:
For the heck of it, I decided to make this test case a little more thorough and add the `-nan` case. And I was slightly surprised that this assertion insisted that there be a minus sign in front of this NaN value. I think maybe this helper is generating the compiled output to the text format, too, and then checking the output matches? Or maybe it's doing some other equality test on the output compared to the result of unmarshaling this string? In any event, perhaps something under the hood is directly examining the sign bit to decide whether this needs to have a sign.

(I didn't dig into the code, but I'd be a little surprised if `< 0` returned true for any NaN, regardless of the sign bit 🤷. And I'd also be surprised if any comparisons of two NaN values _ever_ returned "equal", whether they have sign bits set or not, unless something were actually comparing their bits directly?)

Reply:
I think this makes sense, because the `default_value` field is documented [here](https://github.com/protocolbuffers/protobuf/blob/7dc243c70a870198b80ba4923e8bb8b251e87564/src/google/protobuf/descriptor.proto#L283) to contain "the original text representation of the value." So it sounds like this field is expected to contain the original text without any interpretation. I haven't tested this but I assume if the default value were specified as 1.000 for example then it wouldn't be simplified down to 1 here.

Reply:
Ah. Interestingly, that comment is a bit misleading since it's not _quite_ true: `string` and `bytes` values are decoded (and then, for `bytes` fields, [re-encoded](https://github.com/protocolbuffers/protobuf/blob/v26.1/src/google/protobuf/compiler/parser.cc#L1433)). And, in the example you give with "1.000" as a value, the precision would not actually be preserved because it is just [formatting the numeric value to string](https://github.com/protocolbuffers/protobuf/blob/v26.1/src/google/protobuf/compiler/parser.cc#L1407-L1409). It just happens to capture the negative sign independently and [store that in the value first](https://github.com/protocolbuffers/protobuf/blob/v26.1/src/google/protobuf/compiler/parser.cc#L1399-L1402).

Anyhow, I guess it makes sense now why this value needs to look this way.